[hpe-design-tokens] Add support for isolated test and production environments (#6330)
grommet/hpe-design-system: packages/hpe-design-tokens/docs/FIGMA_CREATE_TOKEN_FILES.md

# Create Figma Token Files Guide

This guide is for HPE Design System maintainers creating new design token files within Figma. New token files might be for previewing token modifications in test enviroments, staging future design token version releases, and other quality assurance tasks.

## Overview

1. [Duplicate master design token files](#duplicating-master-design-token-files).
2. [Prepare files for synchronization](#file-preparation) to `hpe-design-tokens`.
3. [Link library dependencies](#linking-library-dependencies).

## Duplicating master design token files

Master design token files are the currently published design tokens files within the 
[HPE Design System Figma library](https://www.figma.com/files/815326206297160627/project/277423868?fuid=797158032074964818).

1. In Figma, duplicate the Primitive, Semantic, and Component tokens files. The 

> duplicated files will be created in your Figma "Drafts" directory.

Reviewer: ```suggestion
1. In Figma, duplicate the Primitive, Semantic, and Component tokens files. The duplicated files will be created in the same Figma project (folder) the master design token files are in.
```

Author: Whenever I duplicate a file it is always created in my drafts directory... Why do you get a different behavior?

Reviewer: @halocline  Ah, it must be related to edit access. When I duplicate, it doesn't move it to drafts, but duplicates it within the current project.

Author: "Edit access" makes sense. I will modify to account.

Author: Modified

> 2. Move files to a Figma Team and Project directory of your choice. As best practice, 

Reviewer: ```suggestion
2. Move the duplicated files into a Figma Team and Project directory of your choice. Be careful to move only duplicated files, not the master files. As best practice, 
```

> this Team's access should be limited to invited members only so that they are not mistaken from the published master files.

Reviewer: ```suggestion
2. Move files to a Figma Team and Project directory of your choice.

> Best practice: Use a Team with restricted access (invited members only) so working files are not confused with published master files.
```

Author: Added

> ## File preparation

Reviewer: ```suggestion
## File preparation (recommended)
```

Technically optional, the following file preparation steps are recommended for clear 

> identification and recognition of intent (i.e., non-production, testing/QA/staging/preview version).

Reviewer: ```suggestion
These steps are technically optional but strongly recommended so non-production files are clearly identifiable (testing/QA/staging/preview).
```

For each file:
1. Rename the file so that it contains the intent and version. For example, `Design Tokens - Primitives (Copy)` might become ``Design Tokens - Primitives (v3.0.0-alpha)`.
2. Alter the file's "Cover" so that it is visually distinct and recognizable. For example, update the version number to match the file name and alter the background color.

## Linking library dependencies

> At this point, the duplicated working files still contain library references to the master files. These references need to be replaced with references to the equivalent working file.

Reviewer: ```suggestion
After duplication, working files still reference master libraries. These references need to be replaced with references to the equivalent working file.
```

### Primitives working file

1. Open "Manage libraries" from the Assets menu.
2. Remove the master "Design Tokens - Semantic" library.

> 3. Publish the primitive working file. Ensure the publishing scope is limited to the intended team -- not to pan-HPE design teams.

Reviewer: ```suggestion
3. Publish the primitive working file. Set the publishing scope to the intended team only, and **do not** publish to “everyone at HPE (organization)” (pan-HPE).
```

### Semantic working file

1. Open "Manage libraries" from the Assets menu.

> 2. Add "Design Tokens - Primitives (your-working-file's-name)".

Reviewer: ```suggestion
2. Add **Design Tokens - Primitives (your duplicated/new working file name)**..
```

> 3. Replace references to the master primitive file with the working file.

Reviewer: ```suggestion
3. Replace references to the master primitives file with the working primitives file.
```

  1. Note: The master file must be present in order for reference replacement. If the master file is not currently present, add it now.
  2. Select the master "Design Tokents - Primitives".
  3. Select "Swap library".
  4. From "Choose library", select "Design Tokens - Primitives (your-working-file's-name)".
  5. From the "New variable collection," select the like-named equivalent for each "Used variable collection." In this case `primitives` is the only applicable collection.

>   6. Swap the libraries by clicking the "Swap library" button.

Reviewer: ```suggestion
   1. **Note:** The master primitive file must be present to perform a library swap. If it is not present, add it temporarily.
   2. Select the master **Design Tokens - Primitives**.
   3. Select **Swap library**.
   4. From **Choose library**, select **Design Tokens - Primitives (your duplicated/new working file name)**.
   5. In **New variable collection**, map each collection to its like-named equivalent in **Used variable collection**. In this case, only `primitives` applies.
   6. Click **Swap library** to confirm.
```

4. Reopen "Manage libraries".

> 5. Remove the primitives and semantic master files.

Reviewer: ```suggestion
5. Remove the master Design Tokens - Primitives and master Design Tokens - Semantic libraries.
```

> 6. Publish the semantic working file. Ensure the publishing scope is limited to the intended team -- not to pan-HPE design teams.

Reviewer: ```suggestion
6. Publish the semantic working file. Set publishing scope to the intended team only. **Do not** publish to everyone at HPE (organization).
```

  - "Design Tokens - Primitives"
  - "Design Tokens - Primitives (your-working-file's-name)"
4. Replace references to the master files with the working files.
  1. Select the master "Design Tokents - Primitives".
  2. Select "Swap library".
  3. From "Choose library", select "Design Tokens - Primitives (your-working-file's-name)".
  4. From the "New variable collection," select the like-named equivalent for each "Used variable collection." In this case `primitives` is the only applicable collection.
  5. Swap the libraries by clicking the "Swap library" button.
  6. Reopen "Manage libraries".
  7. Select the master "Design Tokents - Semantic".
  8. Select "Swap library".
  9. From "Choose library", select "Design Tokens - Semantic (your-working-file's-name)".
  10. From the "New variable collection," select the like-named equivalent for each "Used variable collection." In this case `color`, `dimension`, and `global`. If all three collections are not present, see [#troubleshooting] before proceeding.
  11. Swap the libraries by clicking the "Swap library" button.
5. Reopen "Manage libraries".
6. Remove the primitives and semantic master files.

> 7. Publish the component working file. Ensure the publishing scope is limited to the intended team -- not to pan-HPE design teams.

Reviewer: ```suggestion
1. Open **Manage libraries** from the Assets menu.
2. Add **Design Tokens - Semantic (your duplicated/new working file name)**.
3. Pending further process validation, also add:
   - **Design Tokens - Primitives**
   - **Design Tokens - Primitives (your duplicated/new working file name)**
4. Replace references to master libraries with working-file libraries.
   1. Select the master **Design Tokens - Primitives**.
   2. Select **Swap library**.
   3. From **Choose library**, select **Design Tokens - Primitives (your duplicated/new working file name)**.
   4. In **New variable collection**, map each collection to its like-named equivalent in **Used variable collection**. In this case, only `primitives` applies.
   5. Click **Swap library**.
5. Reopen **Manage libraries**.
6. Select the master **Design Tokens - Semantic**.
7. Select **Swap library**.
8. From **Choose library**, select **Design Tokens - Semantic (your duplicated/new working file name)**.
9. In **New variable collection**, map each collection to its like-named equivalent in **Used variable collection**. In this case: `color`, `dimension`, and `global`.
   - If all three are not present, see [Troubleshooting](#troubleshooting) before continuing.
10. Click **Swap library**.
11. Reopen **Manage libraries**.
12. Remove the master **Design Tokens - Primitives** and master **Design Tokens - Semantic** libraries.
13. Publish the component working file. Set publishing scope to the intended team only; **do not** publish to **Everyone at HPE (organization)**.
```

> ## Primitives working file, again ☹️

Reviewer: @halocline Is this swapping actual Figma variable references (e.g. primitive -> semantic master reference is replaced with Semantic [new file name] reference, or is it only swapping variables actually used/applied to the file's cover image? (not the variables data, but the applied variables?)

Author: In this instance, it is "swapping variables actually used/applied to the file's cover image."

There are 4 semantic colors which are used in the cover. Those instances are swapped.

  3. From the "New variable collection," select the like-named equivalent for each "Used variable collection." In this case `color` is the only applicable collection.
  4. Swap the libraries by clicking the "Swap library" button.
6. Reopen "Manage libraries".
7. Remove the semantic master file.
  
## Conclusion

The working tokens files have now been duplicated and their requisite library dependencies have be re-linked so that variable collection references are properly mapped within the working files.

Syncing, modifying, and testing with the `sync-tokens-to-figma` and `sync-figma-to-tokens` operational workflows can now be conducted.

## Troubleshooting

Variable collections may only be swapped with collections consumed in the file. If at least one variable from a collection is not used within the file, the collection will not be available for swap (there's nothing available to be swapped out!!!).

As of this writing (2026-07-02), the master "Design Tokens - Component" file does not consume any variables from the semantic/dimension variable collection. To remedy, apply at least one semantic/dimension variable -- from the master file -- to an element within the file.

> Suggested remedy: Select the file's "Cover" layer, alter the cover's corner radius to consume a radius token from the **master** semantic file. Resume libray swap procedure. The `dimension` variable collection should now be available to swap.

Reviewer: ```suggestion
Variable collections can be swapped only when they are actively consumed in the file. If a collection does not appear as swappable, it usually means no variable from that collection is currently in use.

To fix this:

1. Apply at least one variable from the corresponding master library collection to any element in the file.
2. For example, apply a token such as corner radius, color, spacing, or typography to the file's cover image.
3. Reopen **Manage libraries** and retry **Swap library**.

Once the collection is consumed in the file, it should appear as available for swapping.
```

Reviewer: Matt, having added the dimension tokens to the covers, have rewritten this in a way that highlights the possible issue with regards to any collection, and not only the dimension collection.

## Future enhancements

> This procedure includes steps which account for the current state of the master files. Some modifications / file cleanups to the master files (e.g., Use dimension token in component's file cover, color use in primitives' cover) can eliminate some of these steps.

Reviewer: ```suggestion
```

   - **Design Tokens - Primitives (your duplicated/new working file name)**
4. Replace references to master libraries with working-file libraries.
   1. Select the master **Design Tokens - Primitives**.
   2. Select **Swap library**.
   3. From **Choose library**, select **Design Tokens - Primitives (your duplicated/new working file name)**.
   4. In **New variable collection**, map each collection to its like-named equivalent in **Used variable collection**. In this case, only `primitives` applies.
   5. Click **Swap library**.
5. Reopen **Manage libraries**.
6. Select the master **Design Tokens - Semantic**.
7. Select **Swap library**.
8. From **Choose library**, select **Design Tokens - Semantic (your duplicated/new working file name)**.
9. In **New variable collection**, map each collection to its like-named equivalent in **Used variable collection**. In this case: `color`, `dimension`, and `global`.
   - If all three are not present, see [Troubleshooting](#troubleshooting) before continuing.
10. Click **Swap library**.
11. Reopen **Manage libraries**.
12. Remove the master **Design Tokens - Primitives** and master **Design Tokens - Semantic** libraries.

> 13. Publish the component working file. Set publishing scope to the intended team only; **do not** publish to **Everyone at HPE (organization)**.

Reviewer: ```suggestion
1. Open **Manage libraries** from the Assets menu.
2. Add **Design Tokens - Semantic (your duplicated/new working file name)**.
3. Pending further process validation, also add:
   - **Design Tokens - Primitives**
   - **Design Tokens - Primitives (your duplicated/new working file name)**
4. Replace references to master libraries with working-file libraries:
   1. **Swap Primitives**
      1. Select the master **Design Tokens - Primitives**.
      2. Select **Swap library**.
      3. From **Choose library**, select **Design Tokens - Primitives (your duplicated/new working file name)**.
      4. In **New variable collection**, map each collection to its like-named equivalent in **Used variable collection** (typically `primitives`).
      5. Click **Swap library**.
   2. **Swap Semantic**
      1. Reopen **Manage libraries**.
      2. Select the master **Design Tokens - Semantic**.
      3. Select **Swap library**.
      4. From **Choose library**, select **Design Tokens - Semantic (your duplicated/new working file name)**.
      5. In **New variable collection**, map each collection to its like-named equivalent in **Used variable collection**.
      6. If expected collections are missing, see [Troubleshooting](#troubleshooting).
      7. Click **Swap library**.
5. Reopen **Manage libraries**.
6. Remove the master **Design Tokens - Primitives** and master **Design Tokens - Semantic** libraries.
7. Publish the component working file. Set publishing scope to the intended team only; **do not** publish to **Everyone at HPE (organization)**.
```

Reviewer: @halocline I hadn't properly grouped the sub tasks here (4.1-swapping primitives and 4.2 - swapping semantic), now both primitive and semantic have their own numbered groups, making it look more organised.

3. Publish the primitive working file. Set the publishing scope to the intended team only, and **do not** publish to “everyone at HPE (organization)” (pan-HPE).

### Semantic working file

1. Open "Manage libraries" from the Assets menu.
2. Add **Design Tokens - Primitives (your duplicated/new working file name)**..
3. Replace references to the master primitives file with the working primitives file.
   1. **Note:** The master primitive file must be present to perform a library swap. If it is not present, add it temporarily.
   2. Select the master **Design Tokens - Primitives**.
   3. Select **Swap library**.
   4. From **Choose library**, select **Design Tokens - Primitives (your duplicated/new working file name)**.
   5. In **New variable collection**, map each collection to its like-named equivalent in **Used variable collection**. In this case, only `primitives` applies.
   6. Click **Swap library** to confirm.
4. Reopen "Manage libraries".
5. Remove the master Design Tokens - Primitives and master Design Tokens - Semantic libraries.
6. Publish the semantic working file. Set publishing scope to the intended team only. **Do not** publish to everyone at HPE (organization).

> 
> ### Component working file
> 
> 1. Open **Manage libraries** from the Assets menu.
> 2. Add **Design Tokens - Semantic (your duplicated/new working file name)**.
> 3. Pending further process validation, also add:
>    - **Design Tokens - Primitives**
>    - **Design Tokens - Primitives (your duplicated/new working file name)**
> 4. Replace references to master libraries with working-file libraries:
>    1. **Swap Primitives**
>       1. Select the master **Design Tokens - Primitives**.
>       2. Select **Swap library**.
>       3. From **Choose library**, select **Design Tokens - Primitives (your duplicated/new working file name)**.
>       4. In **New variable collection**, map each collection to its like-named equivalent in **Used variable collection** (typically `primitives`).
>       5. Click **Swap library**.
>    2. **Swap Semantic**
>       1. Reopen **Manage libraries**.
>       2. Select the master **Design Tokens - Semantic**.
>       3. Select **Swap library**.
>       4. From **Choose library**, select **Design Tokens - Semantic (your duplicated/new working file name)**.
>       5. In **New variable collection**, map each collection to its like-named equivalent in **Used variable collection**.
>       6. If expected collections are missing, see [Troubleshooting](#troubleshooting).
>       7. Click **Swap library**.
> 5. Reopen **Manage libraries**.
> 6. Remove the master **Design Tokens - Primitives** and master **Design Tokens - Semantic** libraries.
> 7. Publish the component working file. Set publishing scope to the intended team only; **do not** publish to **Everyone at HPE (organization)**.
> 
> ## Primitives working file, again ☹️
> 
> This is a step needed to resolve semantic colors consumed in the primitives' file cover -- a bit of a circular dependency.
> 
> 1. Open "Manage libraries" from the Assets menu.
> 2. Add "Design Tokens - Semantic" master file.
> 3. Add "Design Tokens - Semantic (your-working-file's-name)".
> 4. Replace references to the master files with the working files.
> 5. Select the master "Design Tokents - Semantic".
>   1. Select "Swap library".
>   2. From "Choose library", select "Design Tokens - Semantic (your-working-file's-name)".
>   3. From the "New variable collection," select the like-named equivalent for each "Used variable collection." In this case `color` is the only applicable collection.
>   4. Swap the libraries by clicking the "Swap library" button.
> 6. Reopen "Manage libraries".
> 7. Remove the semantic master file.
>   
> ## Conclusion
> 
> The working tokens files have now been duplicated and their requisite library dependencies have be re-linked so that variable collection references are properly mapped within the working files.
> 
> Syncing, modifying, and testing with the `sync-tokens-to-figma` and `sync-figma-to-tokens` operational workflows can now be conducted.
> 
> ## Troubleshooting
> 
> Variable collections can be swapped only when they are actively consumed in the file. If a collection does not appear as swappable, it usually means no variable from that collection is currently in use.
> 
> To fix this:
> 
> 1. Apply at least one variable from the corresponding master library collection to any element in the file.
> 2. For example, apply a token such as corner radius, color, spacing, or typography to the file's cover image.
> 3. Reopen **Manage libraries** and retry **Swap library**.
> 
> Once the collection is consumed in the file, it should appear as available for swapping.

Reviewer: ```suggestion
# Create Figma Token Files Guide

This guide is for HPE Design System maintainers creating new design token files within Figma. New token files might be used for previewing token modifications in test environments, staging future design token version releases, and other quality assurance tasks.

## Overview

1. [Duplicate master design token files](#duplicating-master-design-token-files).
2. [Prepare files for synchronization](#file-preparation-recommended) to `hpe-design-tokens`.
3. [Link library dependencies](#linking-library-dependencies).

## Duplicating master design token files

Master design token files are the currently published design token files within the [HPE Design System Figma library](https://www.figma.com/files/815326206297160627/project/277423868?fuid=797158032074964818).

1. In Figma, duplicate the Primitive, Semantic, and Component token files.  
   The duplicated files will be created in either the same project folder as the master design token files or in your Figma **Drafts** directory. The location depends on your **edit permissions** for the project.
2. Move the duplicated files into a Figma Team and Project directory of your choice. Be careful to move only duplicated files, not the master files.

> **Best practice:** Use a Team with restricted access (invited members only) so working files are not confused with published master files.

## File preparation (recommended)

These steps are technically optional but strongly recommended so non-production files are clearly identifiable (testing/QA/staging/preview).

For each file:

1. Rename the file so that it contains the intent and version.  
   For example, `Design Tokens - Primitives (Copy)` might become `Design Tokens - Primitives (v3.0.0-alpha)`.
2. Alter the file’s Cover so that it is visually distinct and recognizable.  
   For example, update the version number to match the file name and alter the background color.

## Linking library dependencies

After duplication, working files still reference master libraries. These references must be replaced with references to the equivalent working files.

### Primitives working file

1. Open **Manage libraries** from the Assets menu.
2. Remove the master `Design Tokens - Semantic` library.
3. Publish the primitive working file. Set the publishing scope to the intended team only, and **do not** publish to **Everyone at HPE (organization)** (pan-HPE).

### Semantic working file

1. Open **Manage libraries** from the Assets menu.
2. Add `Design Tokens - Primitives (your duplicated/new working file name)`.
3. Replace references to the master primitives file with the working primitives file.
   1. **Note:** The master primitive file must be present to perform a library swap. If it is not present, add it temporarily.
   2. Select the master `Design Tokens - Primitives`.
   3. Select **Swap library**.
   4. From **Choose library**, select `Design Tokens - Primitives (your duplicated/new working file name)`.
   5. In **New variable collection**, map each collection to its like-named equivalent in **Used variable collection**. In this case, only `primitives` applies.
   6. Click **Swap library** to confirm.
4. Reopen **Manage libraries**.
5. Remove the master `Design Tokens - Primitives` and master `Design Tokens - Semantic` libraries.
6. Publish the semantic working file. Set publishing scope to the intended team only, and **do not** publish to **Everyone at HPE (organization)**.

### Component working file

1. Open **Manage libraries** from the Assets menu.
2. Add `Design Tokens - Semantic (your duplicated/new working file name)`.
3. Pending further process validation, also add:
   - `Design Tokens - Primitives`
   - `Design Tokens - Primitives (your duplicated/new working file name)`
4. Replace references to master libraries with working-file libraries:
   1. **Swap Primitives**
      1. Select the master `Design Tokens - Primitives`.
      2. Select **Swap library**.
      3. From **Choose library**, select `Design Tokens - Primitives (your duplicated/new working file name)`.
      4. In **New variable collection**, map each collection to its like-named equivalent in **Used variable collection** (typically `primitives`).
      5. Click **Swap library**.
   2. **Swap Semantic**
      1. Reopen **Manage libraries**.
      2. Select the master `Design Tokens - Semantic`.
      3. Select **Swap library**.
      4. From **Choose library**, select `Design Tokens - Semantic (your duplicated/new working file name)`.
      5. In **New variable collection**, map each collection to its like-named equivalent in **Used variable collection**.
      6. If expected collections are missing, see [Troubleshooting](#troubleshooting).
      7. Click **Swap library**.
5. Reopen **Manage libraries**.
6. Remove the master `Design Tokens - Primitives` and master `Design Tokens - Semantic` libraries.
7. Publish the component working file. Set publishing scope to the intended team only, and **do not** publish to **Everyone at HPE (organization)**.

### Primitives working file (semantic cover dependency)

This step resolves semantic color references consumed by the Primitives file cover (a circular dependency).

1. Open **Manage libraries** from the Assets menu.
2. Add the master `Design Tokens - Semantic` library.
3. Add `Design Tokens - Semantic (your working file name)`.
4. Replace references to the master semantic library with the working semantic library:
   1. Select the master `Design Tokens - Semantic`.
   2. Select **Swap library**.
   3. From **Choose library**, select `Design Tokens - Semantic (your working file name)`.
   4. In **New variable collection**, map each collection to its like-named equivalent in **Used variable collection**. In this case, only `color` applies.
   5. Click **Swap library**.
5. Reopen **Manage libraries**.
6. Remove the master `Design Tokens - Semantic` library.

## Conclusion

The working token files have now been duplicated and their required library dependencies relinked so that variable collection references are properly mapped within the working files.

Syncing, modifying, and testing with the `sync-tokens-to-figma` and `sync-figma-to-tokens` operational workflows can now be conducted.

## Troubleshooting

Variable collections can be swapped only when they are actively consumed in the file. If a collection does not appear as swappable, it usually means no variable from that collection is currently in use.

To fix this:

1. Apply at least one variable from the corresponding master library collection to any element in the file.
2. For example, apply a token such as corner radius, color, spacing, or typography to the file’s cover image.
3. Reopen **Manage libraries** and retry **Swap library**.

Once the collection is consumed in the file, it should appear as available for swapping.
```
